Enhance: allow wait for more network packets to be adjusted

SlySven opened the pull request:
This might be useful for tweaking for MUDs that do not use Go Aheads for end-users with a good network connection.

This commit does ~~not currently~~ save the (per profile) setting ~~but if it proves useful it might need that as well.~~ *done in the subsequent 5529478a59a7fc5374426e382a089d78ecc3e582 commit...*

This PR also performs a maintenance reordering of items in any `QGridLayout` in the `src/ui/profile_preferences.ui` so that they are back in ascending row/column order again.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: allow wait for more network packets to be adjusted

This might be useful for tweaking for MUDs that do not use Go Aheads for
end-users with a good network connection.

This commit does not currently save the (per profile) setting but if it
proves useful it might need that as well.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
The new setting is "Special Options" -> "Other Special Options" as "Network packet timeout:":
![image](https://user-images.githubusercontent.com/6163092/131266725-62676432-d414-48c4-88fb-49e4dd136809.png)

SlySven commented:
I've only marked this as "Ready for review" so as to get the CI builds to proceed (it didn't seem to produce the three OS compiled "builds" as a draft for me)...

Or maybe I need that "refresh links" meta command or whatever it is?

SlySven commented:
/refresh links

vadi2 commented:
Agree we need to do something about this. Would love to hear others thought about it

Eraene commented:
I like this option. While you might run the risk of people tinkering with it and then getting undesired results and not understanding why, I think the net improvement is worth it.

vadi2 commented:
/refresh links

vadi2 commented:
I'm not sure what happned the the links bot - I'll close and re-open the PR so it can re-create it.

SlySven committed:
Revise: following peer review of a duplicate PR

In #5415 which is a (presumably accidental) duplicate of this PR @vadi2
suggested the following as a replacement for the tool-tip for the option:
"Go-Ahead (GA) tells Mudlet when the game server is done sending text. On
games that don't support GA, this option controls how long Mudlet will wait
for more text to arrive. Larger values will help reduce the risk that a
large piece of text has unintended linebreaks in the middle of it, which
would break triggers. Lesser values will increase the risk of text getting
broken up, but will make the game feel more responsive."

in case it is not clear from the commit I have revised the text to partial
include some of that - and also taken on board Leris/Kebap who pointed out
on Discord that GA is not the only means of the Server signalling that it
has finished sending out packets. The text is now along the lines of
(within Markup limitations):
"*Go-Ahead* (`GA`) / *End-of-record* (`EOR`) signalling are Telnet enhancements
that tell Mudlet when the Game Server is done sending a piece of text. On
Game Servers which do not provide `GA` or `EOR` this option controls how
long Mudlet will wait for further network packets (i.e. text or other
'out-of-band' `OOB` data) to arrive. The default is 300 milli-seconds.
Larger values will help to prevent unintended linebreaks in the middle of
big pieces of text which may break triggers; smaller values may make Mudlet
seem more responsive to incoming text but runs a risk of breaking up text
(or more problematically `OOB` data).

*Adjustment of this persistent per-profile control is **NOT** recommended
unless you understand what it is doing and why it might be helpful.
Fine-tuning is likely to depend on particular network conditions and MUD
Game servers and a lower (shorter wait) setting that works for one user may
not be long enough for others.*"

He also suggested replacing the option as it is called
"Network packet timeout:"

on the basis that: "This text is great for an engineer who knows the
context, but not so much for a player who doesn't. How about:
'Wait up to [ ]ms for more text to arrive?'"

however I felt that was too wordy and could get broken in translation to
other locales. Also - given the potential for breakage,
particularly if it is set too small - we do not want the player who does
not get it, to play around with it.

Note that although I now mention something to the effect that the setting
is per profile and is saved between sessions that has not been coded yet.
It will follow in an additional commit that is pending.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
@SlySven there is an explanation here:

> I'm not sure what happned the the links bot - I'll close and re-open the PR so it can re-create it.

SlySven committed:
Enhance: save Network Packet Timeout between profile sessions

Also revise the tool-tip text and move the display of the unit (mSec) to
the label from the spin-box - to make translation a tiny bit simpler.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Well, if you really mean to use:>
> Wait up to [ ]ms for more text to arrive?

do you mean you want to put the spin box in the middle of two labels? That might look nice for the en locale but it would not be good **I18n** (or perhaps that is **L10n**)! OTOH If you mean that as a single label for the UI that still looks to me to be a bit on the long side - and likely it would be even longer for some other locales.

Is there any other suggestions besides that (and the current `Network packet timeout (mSec):`) that anyone can come up with?

vadi2 commented:
UI: 

> Wait up to [ ]s for more text to arrive

Note the change to seconds - people don't typically think in terms of miliseconds, so `0.3` would be better, and they can adjust it to `0.03` if they'd like.

Regarding i18n: let's wait until it's an actual problem, right now it's theoretical. I think if we put the entire message in the context and add screenshots of how it looks like to Crowdin, that'll be OK.

Tooltip:

> Go-Ahead (GA) or EOR tells Mudlet when the game server is done sending text. On games that don't support GA, this option controls how long Mudlet will wait for more text to arrive. Larger values will help reduce the risk that a large piece of text has unintended linebreaks in the middle of it, which would break triggers. Lesser values will increase the risk of text getting broken up, but will make the game feel more responsive.

SlySven commented:
> Note the change to seconds - people don't typically think in terms of miliseconds, so 0.3 would be better, and they can adjust it to 0.03 if they'd like.

I was using a `QSpinBox` using an `int` with a `10` step size (so they are not too small to get from 10 to 500) - I'd have to check how the `QDoubleSpinBox` which uses a `double` if/how it can be configure for A) only 3 decimal places and B) a step size of `0.01`...

> Go-Ahead (GA) or EOR tells Mudlet when the game server is done sending text. On games that don't support GA, this option controls how long Mudlet will wait for more text to arrive. Larger values will help reduce the risk that a large piece of text has unintended linebreaks in the middle of it, which would break triggers. Lesser values will increase the risk of text getting broken up, but will make the game feel more responsive.

I think to protect the novice end-user we have to include a setence to indicate what the previous default value (0.300 S) was so that they can reset it if they frob the setting too much so it breaks for their situation...

vadi2 commented:
Agreed, that's a good point.

Kebap commented:
My suggestion walks this way:
> Wait how long for more text to arrive?

The "seconds" bit is already in the widget text.

Agree with @SlySven here, we shouldn't put a widget in the middle of a label sentence.

Also thanks for briefly mentioning End-Of-Record as a GA alternative 👍

vadi2 commented:
That would work as well 👍

SlySven commented:
i intend to revise the tool-tip in the next commit to:
><p><i>Go-Ahead</i> (<tt>GA</tt>) and <i>End-of-record</i> (<tt>EOR</tt>) signalling tells Mudlet when the game server is done sending text. On games that do not provide <tt>GA</tt> or <tt>EOR</tt>, this option controls how long Mudlet will wait for more text to arrive. Greater values will help reduce the risk that Mudlet will split a large piece of text (with unintended line-breaks in the middle) which can stop some triggers from working. Lesser values increases the risk of text getting broken up, but may make the game feel more responsive.</p><p><i>The default value, which was what Mudlet used before this control was added, is 0.300 Seconds.</i></p>

In comparison to the suggested:
>~Go-Ahead (GA) or~**1**~EOR tells Mudlet when the game server is done sending text. On games that don't support GA~**2**~, this option controls how long Mudlet will wait for more text to arrive. Larger~**3** ~values will help reduce the risk that a large piece of text has unintended linebreaks in the middle of it, which would break~**4** ~triggers. Lesser values will increase the risk of text getting broken up, but will~**5** ~make the game feel more responsive.~

1. They are not exclusive so "and" is better than "or" i believe - and it is fair to describe them as *signalling* methods - which might not be obvious to the novice user.
2. For consistency we should mention both again here.
3. *Greater* is a better anonym than Larger to *Lesser*.
4. Given that we have mentioned line-**breaks** in the sentence already, using the word "break" in a different (failure) context  is perhaps not the wisest word choice here.
5. I don't want to promise something with a "will" - I feel a "may" is less assertive here! :stuck_out_tongue_winking_eye:

SlySven committed:
Revise: label and tool-tip for the new Network Packet Timout knob

Also:
* To allow for the control to display the time in seconds it has to be
converted from a `QSpinBox` to a `QDoubleSpinBox` and some properties
tweaked to match - beyond the control the C++ code still uses an integer
representing the time out in milliseconds.
* Reorder the tab order for a couple of other controls that have been
added since it was last redone - otherwise tabbing across them will cause
odd jumps between different pages of the profile preferences dialogue.
* Remove a redundant default text for a label that "lights up" when
profile passwords are being migrated between secure and portable storage
modes - the text should never shown to the end-user - but just in case it
does a slightly more informative message than the default translatable
"TextLabel" will be used.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> Wait how long for more text to arrive?

Since everyone else seems to be rejecting a perfectly precise label in favour of something that is supposedly more *user-friendly* :roll_eyes: I intend to use a shorter, and which also does not include a `?` so is better suited to be a label, text:
`Additional text wait time:`

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -3479,11 +3482,46 @@ you can use it but there could be issues with aligning columns of text</string>
+           <widget class="QDoubleSpinBox" name="doubleSpinBox_networkPacketTimeout">
             <property name="toolTip">
-             <string>&lt;p&gt;If checked this will cause all problem Unicode codepoints to be reported in the debug output as they occur; if cleared then each different one will only be reported once and summarized in as a table when the console in which they occurred is finally destroyed (when the profile is closed).&lt;/p&gt;</string>
+             <string>&lt;p&gt;&lt;i&gt;Go-Ahead&lt;/i&gt; (&lt;tt&gt;GA&lt;/tt&gt;) and &lt;i&gt;End-of-record&lt;/i&gt; (&lt;tt&gt;EOR&lt;/tt&gt;) signalling tells Mudlet when the game server is done sending text. On games that do not provide &lt;tt&gt;GA&lt;/tt&gt; or &lt;tt&gt;EOR&lt;/tt&gt;, this option controls how long Mudlet will wait for more text to arrive. Greater values will help reduc ...
```
```suggestion
             <string>&lt;p&gt;&lt;i&gt;Go-Ahead&lt;/i&gt; (&lt;tt&gt;GA&lt;/tt&gt;) and &lt;i&gt;End-of-record&lt;/i&gt; (&lt;tt&gt;EOR&lt;/tt&gt;) signalling tells Mudlet when the game server is done sending text. On games that do not provide &lt;tt&gt;GA&lt;/tt&gt; or &lt;tt&gt;EOR&lt;/tt&gt;, this option controls how long Mudlet will wait for more text to arrive. Greater values will help reduce the risk that Mudlet will split a large piece of text (with unintended line-breaks in the middle) which can stop some triggers from working. Lesser values increases the risk of text getting broken up, but may make the game feel more responsive.&lt;/p&gt;&lt;p&gt;&lt;i&gt;The default value is 0.300 seconds.&lt;/i&gt;&lt;/p&gt;</string>
```

SlySven replied:
I'd rather not loose the reference to how things were because it does explain why we have chosen that as a default - there is not a clear explanation as to the reason for that figure otherwise - just Heiko's commit comment in cf6adfebbe612a0671b2074993a4ed3b9e8ccf71 back in 2011:
> bugfix: logon screens are now properly displayed

vadi2 replied:
There's no need for this historical information as far as players are concerned, though.

vadi2 commented on src/ui/profile_preferences.ui:
```diff
@@ -3479,11 +3482,46 @@ you can use it but there could be issues with aligning columns of text</string>
+             <string>&lt;p&gt;&lt;i&gt;Go-Ahead&lt;/i&gt; (&lt;tt&gt;GA&lt;/tt&gt;) and &lt;i&gt;End-of-record&lt;/i&gt; (&lt;tt&gt;EOR&lt;/tt&gt;) signalling tells Mudlet when the game server is done sending text. On games that do not provide &lt;tt&gt;GA&lt;/tt&gt; or &lt;tt&gt;EOR&lt;/tt&gt;, this option controls how long Mudlet will wait for more text to arrive. Greater values will help reduc ...
+            </property>
+            <property name="suffix">
+             <string extracomment="For most locales a space should be included so that the text is separated from the number!"> Seconds</string>
```
```suggestion
             <string extracomment="For most locales a space should be included so that the text is separated from the number!"> seconds</string>
```

SlySven replied:
I was confused/mistaken and mixing this up with using `S` when  the unit is abbreviated.

vadi2 reviewed (commented):
I set the option at 0.01 but it seems to be waiting longer. I take it that's just the minimum time between two network packets for me? Is everyone who plays wotmud dealing with this big delay?

![image](https://user-images.githubusercontent.com/110988/134796536-8270666d-dc13-4aed-be51-5b1dcacb588a.png)

vadi2 commented:
This is getting really close now. Could you also make it so the option is disabled if you're playing a game that has GA enabled? That way people who aren't affected by it won't get confused and try to tweak the option unnecessarily. (and we will be getting people like this)

Kebap commented:
> everyone else seems to be rejecting a perfectly precise label in favour of something that is supposedly more _user-friendly_ 🙄 

Are you meaning "network package timeout" is perfectly precise?
I mean maybe but how understandable is it if you want to play a game and have no clue about network technology underneath?

> I intend to use a shorter, and which also does not include a `?` so is better suited to be a label, text:
> `Additional text wait time:`

Well to me that now seems neither precise nor understandable.

I asked a friend to read your label and tool-tip and guess what it toggles. They are very IT affine and indeed know the internet works with packages. They failed to grasp the meaning. The tool-tip was so long, they stopped reading after 2-3 lines, not knowing what "Go ahead" and those concepts at start all refer to.
Then I showed them the other suggestion "Wait how long for more text to arrive?" and they concluded: "Oh, so this is how long I have to wait for answer after I send a command to game?"
That seems fairly on point, only difference is the game could also send text without prior commands, but that's a good example nonetheless.
So here is my next suggestion (without question mark as well):
"Wait at max this long to show game text:"

As I said, the tool-tip also needs some work. Let's start with a general description of what it's about, and not these minuscule telnet specifications that players may or may not even know about.

```
How long should Mudlet wait until the game server is done sending text? Sometimes so much text is sent, it won't fit all into one package. If the game does not signal the end with Go-Ahead (GA) and End-of-record (EOR), Mudlet will not show the text immediately, as it may be incomplete. 
This option controls how long Mudlet will wait for more text to arrive. Greater values will help reduce the risk that Mudlet will split a large piece of text (with unintended line-breaks in the middle) which can stop some triggers from working. Lesser values increases the risk of text getting broken up, but may make the game feel more responsive.
The default value, which was what Mudlet used before this control was added, is 0.300 Seconds.
```

That said, I really like the functionality, should have started by saying that.

vadi2 commented:
Isn't your tooltip suggestion longer?

FTR it was received well in the MUD discord: https://discord.com/channels/279748146316312576/359062772337606656/891573051794288671

Kebap commented:
> Isn't your tooltip suggestion longer?

Yeah I added additional explanation in the beginning and cycled other sentences around.
Would need to spend more time thinking about better text formulation. Writing precise and understandable and short at the same time is no easy task. That's no good reason to stop a precise however.

> FTR it was received well in the MUD discord

The feature? Sure.
The understandability of the tool-tip? Seems a bit tainted as you gave it away in your question already. Even then, one person managed to misunderstand it still.

SlySven commented:
> This is getting really close now. Could you also make it so the option is disabled if you're playing a game that has GA enabled? That way people who aren't affected by it won't get confused and try to tweak the option unnecessarily. (and we will be getting people like this)

All profiles have GA enabled (we do NOT support the "Suppress Go Ahead" Telnet sub-option **3**) - it is just that some do not send them - and doing what you propose seems to me to be making this PR more involved... :worried:

vadi2 commented:
I think that'll be unfortunately necessary because otherwise it'll be adding confusion to those whom it's not related to.

SlySven committed:
Update src/ui/profile_preferences.ui

Correct miss-capitalisation of time unit in control suffix.

Co-authored-by: Vadim Peretokin <vperetokin@gmail.com>

SlySven commented:
> I think that'll be unfortunately necessary because otherwise it'll be adding confusion to those whom it's not related to.

:thinking: Well the timer is still present and active when GAs and/or EORs are in the system - they just short-circuit the timer so it never gets to have an effect if I remember the coding correctly. Furthermore if such extra code to do this was somehow put together and the profile was loaded offline (without making a connection) there would not be those GA/EORs to "disable" the control and so the user could still frob it then. I am not certain that we need the extra complexity - I mean the current tool-tip even says:
> ...On games that do not provide GA or EOR, this option controls how long Mudlet will wait for more text to arrive...

fair enough we do not explicitly say "so if your Game server IS using them you don't need to fiddle with this control as it will not have any effect" but can we not let them work that out for themselves?

SlySven commented:
@Kebap If you can come up with some concise alternative texts then we can revisit this I guess. I do know there are a few (WoTMUD) users who are eagerly looking forward to this feature... :drooling_face:

vadi2 commented:
https://github.com/Mudlet/Mudlet/pull/5403#issuecomment-927319031 this is still important - we can't just please one segment while confusing the other, that's not an improvement 😉
